src/state/notifications/sagas.ts

```diff
@@ -68,7 +83,73 @@ export function* setNotificationEmailSaga(action: SetNotificationEmailAction | u
   }
 }
 
+export function* subscribeWalletSaga(action: SubscribeWalletAction) {
+  const { payload } = action as SubscribeWalletAction;
+  try {
+    const response: { session_token: string; result: Result } = yield call(subscribeEmail, payload);
+    if (response.session_token) {
+      yield put(subscribeWalletSuccess(response.session_token));
+    }
+  } catch (error) {
+    yield put(subscribeWalletFailure(error.msg));
+  }
+}
+
+export function* unsubscribeWalletSaga(action: UnsubscribeWalletAction) {
+  const { payload } = action as UnsubscribeWalletAction;
+  try {
+    const response = yield call(unsubscribeEmail, payload);
+    if (response.session_token) {
+      yield put(unsubscribeWalletSuccess(response.session_token));
+    }
+  } catch (error) {
+    yield put(unsubscribeWalletFailure(error.msg));
+  }
+}
+
+export function* authenticateEmailSaga(action: AuthenticateEmailAction) {
+  const { payload, meta } = action as AuthenticateEmailAction;
+  try {
+    const response = yield call(authenticateEmail, payload);
+    if (response.result === Result.success) {
+      yield put(authenticateEmailSuccess());
+      if (meta?.onSuccess) {
+        meta.onSuccess();
+      }
+    }
+  } catch (error) {
+    yield put(authenticateEmailFailure(error.message));
+    if (meta?.onFailure) {
+      meta.onFailure();
+    }
+  }
+}
+
+export function* checkSubscriptionSaga(action: CheckSubscriptionAction) {
+  const { wallets, email } = action.payload;
+  try {
+    const walletWithHashes = yield Promise.all(
+      wallets.map(async wallet => ({ ...wallet, hash: await getWalletHashedPublicKeys(wallet) })),
+    );
+    const hashes = walletWithHashes.map((wallet: Wallet) => wallet.hash);
+    const response = yield call(checkSubscriptionEmail, { hashes, email });
+    const calls: any[] = [];
+    walletWithHashes.map((wallet: Wallet, index: number) => {
```

Comment:
it's hard to read and not an optimal solution. A more optimal way would be to filter all wallets that have successful results from the backend and then call only one action to redux with all of these wallets like this.
` yield put(checkSubscriptionSuccess(subscribedWalletsIds)`
then in the redux instead of 

case NotificationAction.CheckSubscriptionSuccessAction: {
      return {
        ...state,
        subscribedIds: state.subscribedIds.some(id => id === action.payload.result)
          ? action.payload.version === CheckSubscriptionVersion.ADD
            ? [...state.subscribedIds]
            : [...state.subscribedIds].filter(id => id !== action.payload.result)
          : action.payload.version === CheckSubscriptionVersion.ADD
          ? [...state.subscribedIds, action.payload.result]
          : [...state.subscribedIds],
      };
    } 
    
    case NotificationAction.CheckSubscriptionSuccessAction: {
      return {
        ...state,
        subscribedIds: action.payload.subscribedIds
      };
    }

Reply:
Yes, this is not perfect at all, but if I get it right you want to check if any of all our wallets are subscribed every time? So going to the WalletDetails screen you would exam all existing wallets to find out if this particular one is subscribed. Then it's overkill.

Reply:
we send wallets as an argument in action payload, so on wallet details action payload will have only one wallet  [wallet], and only one subscribedIds in the redux. 
Every time when we enter the notifications screen we have to check subscriptions of all wallets anyway, so we send all wallets in an array of check subscription action and all subscribedIds will go into redux

Reply:
done I think